components/script/dom/paintworkletglobalscope.rs

```diff
@@ -54,44 +84,223 @@ impl PaintWorkletGlobalScope {
+    fn invoke_a_paint_callback(&self,
+                               name: Atom,
+                               size: Size2D<Au>,
+                               sender: Sender<Result<Image, PaintWorkletError>>)
+    {
+        let width = size.width.to_px().abs() as u32;
+        let height = size.height.to_px().abs() as u32;
+        debug!("Invoking a paint callback {}({},{}).", name, width, height);
+
+        let cx = self.worklet_global.get_cx();
+        let _ac = JSAutoCompartment::new(cx, self.worklet_global.reflector().get_jsobject().get());
+
+        // TODO: Steps 1-2.1.
+        // Step 3.
+        let definition = match self.paint_definitions.borrow().get(&name) {
+            Some(defn) => defn.clone(),
+            None => {
+                // Step 2.2.
+                warn!("Drawing un-registered paint definition {}.", name);
+                let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                let _ = sender.send(Ok(image));
+                return;
+            }
+        };
+
+        // Steps 4-5
+        rooted!(in(cx) let mut paint_instance = UndefinedValue());
+        match self.paint_class_instances.borrow_mut().entry(name.clone()) {
+            Entry::Occupied(entry) => paint_instance.set(entry.get().get()),
+            Entry::Vacant(entry) => {
+                // Step 5.1
+                if !definition.constructor_valid_flag.get() {
+                    debug!("Drawing invalid paint definition {}.", name);
+                    let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                    let _ = sender.send(Ok(image));
+                    return;
+                }
+                // Step 5.2-5.3
+                let args = HandleValueArray::new();
+                rooted!(in(cx) let mut result = null_mut());
+                unsafe { Construct1(cx, definition.class_constructor.handle(), &args, result.handle_mut()); }
+                paint_instance.set(ObjectValue(result.get()));
+                if unsafe { JS_IsExceptionPending(cx) } {
+                    debug!("Paint constructor threw an exception {}.", name);
+                    unsafe { JS_ClearPendingException(cx); }
+                    definition.constructor_valid_flag.set(false);
+                    let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                    let _ = sender.send(Ok(image));
+                    return;
+                }
+                // Step 5.4
+                entry.insert(Heap::new(paint_instance.get()));
+            }
+        };
+
+        // TODO: Steps 6-7
+        // Step 8
+        let rendering_context = PaintRenderingContext2D::new(self);
+
+        // Step 9
+        let paint_size = PaintSize::new(self, size);
+
+        // TODO: Step 10
+        // Steps 11-12
+        debug!("Invoking paint function {}.", name);
+        let args = unsafe { HandleValueArray::from_rooted_slice(&[
```

Comment:
Well I tried, but I hit:
```
   --> /home/ajeffrey/github/asajeffrey/servo/components/script/dom/macros.rs:549:25
    |
549 |         let mut $name = $crate::dom::bindings::trace::RootedVec::new(&mut root);
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `dom::bindings::trace::JSTraceable` is not implemented for `js::jsapi::Value`
    | 
   ::: /home/ajeffrey/github/asajeffrey/servo/components/script/dom/paintworkletglobalscope.rs
    |
171 |         rooted_vec!(let mut args);
    |         -------------------------- in this macro invocation
    |
    = note: required by `<dom::bindings::trace::RootedVec<'a, T>>::new`
```
The problem being that I need to construct a `RootedVec<JSVal>`.

There's a similar type error with `rooted!(in(cx) let args = ....)`:
```
error[E0277]: the trait bound `js::jsapi::HandleValueArray: js::rust::RootKind` is not satisfied
   --> /home/ajeffrey/github/asajeffrey/servo/components/script/dom/paintworkletglobalscope.rs:171:9
    |
171 | /         rooted!(in(cx) let args = unsafe { HandleValueArray::from_rooted_slice(&[
172 | |             ObjectValue(rendering_context.reflector().get_jsobject().get()),
173 | |             ObjectValue(paint_size.reflector().get_jsobject().get()),
174 | |         ]) });
    | |______________^ the trait `js::rust::RootKind` is not implemented for `js::jsapi::HandleValueArray`
    |
    = note: required by `<js::rust::RootedGuard<'a, T>>::new`
    = note: this error originates in a macro outside of the current crate
```

I'm not sure this is needed anyway, all the objects in the array are already rooted, the only thing is the array itself, but I believe `Call` manages the lifetime of the arguments object?